numpy/core/src/umath/string_ufuncs.cpp

```diff
@@ -0,0 +1,450 @@
+    if (add_loop(umath, "less", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_LE, npy_byte>;
+    if (add_loop(umath, "less_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_GT, npy_byte>;
+    if (add_loop(umath, "greater", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_GE, npy_byte>;
+    if (add_loop(umath, "greater_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+
+    /* All Unicode loops */
+    dtypes[0] = Unicode;
+    dtypes[1] = Unicode;
+
+    loop = string_comparison_loop<false, Py_EQ, npy_ucs4>;
+    if (add_loop(umath, "equal", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_NE, npy_ucs4>;
+    if (add_loop(umath, "not_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_LT, npy_ucs4>;
+    if (add_loop(umath, "less", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_LE, npy_ucs4>;
+    if (add_loop(umath, "less_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_GT, npy_ucs4>;
+    if (add_loop(umath, "greater", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_GE, npy_ucs4>;
+    if (add_loop(umath, "greater_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+
+    res = 0;
+  finish:
+    Py_DECREF(String);
+    Py_DECREF(Unicode);
+    Py_DECREF(Bool);
+    return res;
+}
+
+
+template <bool rstrip, typename character>
+static PyArrayMethod_StridedLoop *
+get_strided_loop(int comp)
+{
+    if (comp == Py_EQ) {
+        return string_comparison_loop<rstrip, Py_EQ, character>;
+    }
+    else if (comp == Py_NE) {
+        return string_comparison_loop<rstrip, Py_NE, character>;
+    }
+    else if (comp == Py_LT) {
+        return string_comparison_loop<rstrip, Py_LT, character>;
+    }
+    else if (comp == Py_LE) {
+        return string_comparison_loop<rstrip, Py_LE, character>;
+    }
+    else if (comp == Py_GT) {
+        return string_comparison_loop<rstrip, Py_GT, character>;
+    }
+    else if (comp == Py_GE) {
+        return string_comparison_loop<rstrip, Py_GE, character>;
+    }
+    assert(0);
+}
+
+
+/*
+ * This function is used for `compare_chararrays` (and void comparisons
+ * currently).  The first could probably be deprecated.
+ *
+ * The `rstrip` mechanism is presumably for some fortran compat, but the
+ * question is whether it would not be better to have/use `rstrip` on such
+ * an array first...
+ *
+ * NOTE: This function is also used for unstructured voids, this works because
+ *       `npy_byte` works for it.
+ */
+extern "C" {
+    NPY_NO_EXPORT PyObject *
+    _umath_strings_richcompare(
+            PyArrayObject *self, PyArrayObject *other, int cmp_op, int rstrip);
+}
```

Comment:
Is this how we do this for C++?  I am not sure which header I could cram this in so that i can just include it from `multiarray/multiarrayobject.c` and `multiarray/arrayobject.c` which need access to these.  Currently I put it there directly, which is OK, but I don't like it too much.

Reply:
Why not in an `string_ufuncs.h` ?